File: src/services/ping.service.ts
/**
 * Ping Service - Perform ping checks
 */
import { exec } from 'child_process';
import { promisify } from 'util';
import { NodeInfo, PingResult } from '../types';
import { resolveHostname } from '../utils/helpers';

const execAsync = promisify(exec);

export class PingService {
  private pingCount = 4;
  private timeout = 3;

  async pingHost(host: string, node: NodeInfo, options?: { count?: number }): Promise<PingResult[][]> {
    const { hostname } = this.parseHost(host);
    const count = options?.count || this.pingCount;
    
    const ip = await resolveHostname(hostname);
    if (!ip) {
      return [[null as any]];
    }

    const results: PingResult[] = [];
    for (let i = 0; i < count; i++) {
      const result = await this.pingIP(ip);
      // Always include IP in result, even if ping failed
      results.push({
        ...result,
        ip: result.ip || ip
      });
    }

    return [results];
  }

  private parseHost(host: string): { hostname: string; port?: number } {
    if (host.includes('://')) {
      host = host.split('://')[1];
    }
    if (host.includes(':')) {
      const parts = host.split(':');
      const port = parseInt(parts[parts.length - 1], 10);
      if (!isNaN(port)) {
        return { hostname: parts.slice(0, -1).join(':') };
      }
    }
    return { hostname: host };
  }

  private async pingIP(ip: string): Promise<PingResult> {
    try {
      const isWindows = process.platform === 'win32';
      const cmd = isWindows
        ? `ping -n 1 -w ${this.timeout * 1000} ${ip}`
        : `ping -c 1 -W ${this.timeout} ${ip}`;

      const { stdout, stderr } = await execAsync(cmd, { timeout: (this.timeout + 1) * 1000 });
      
      // Debug logging (can be removed in production)
      if (stderr && stderr.trim()) {
        console.log(`Ping stderr for ${ip}:`, stderr);
      }
      
      const time = this.parsePingTime(stdout, isWindows);
      
      // Debug logging
      if (time === null) {
        console.log(`Could not parse ping time for ${ip}. Output:`, stdout.substring(0, 200));
      }

      // If we couldn't parse the time, check if ping actually succeeded
      // On Linux, a successful ping will have "1 received" in the output
      // On Windows, a successful ping will have "Reply from" in the output
      if (time === null) {
        const hasSuccess = isWindows 
          ? stdout.includes('Reply from') || stdout.includes('bytes=')
          : stdout.includes('1 received') || stdout.match(/time=/) !== null;
        
        if (!hasSuccess) {
          // Ping command didn't fail, but no successful response
          return {
            status: 'TIMEOUT',
            time: this.timeout,
            ip
          };
        }
        
        // Ping succeeded but we couldn't parse time - try alternative patterns
        // Try to extract time from different formats
        const altMatch = stdout.match(/(\d+\.?\d*)\s*ms/i);
        if (altMatch) {
          const parsedTime = parseFloat(altMatch[1]) / 1000.0;
          return {
            status: 'OK',
            time: parsedTime,
            ip
          };
        }
        
        // If we still can't parse, but ping succeeded, return with 0 time
        // This is better than marking it as failed
        return {
          status: 'OK',
          time: 0,
          ip
        };
      }

      return {
        status: 'OK',
        time: time,
        ip
      };
    } catch (error: any) {
      // Debug logging
      console.log(`Ping error for ${ip}:`, {
        code: error.code,
        signal: error.signal,
        message: error.message,
        stdout: error.stdout?.substring(0, 200),
        stderr: error.stderr?.substring(0, 200)
      });
      
      // Always include IP in result, even if ping failed
      if (error.code === 'ETIMEDOUT' || error.signal === 'SIGTERM' || error.code === 'TIMEOUT') {
        return {
          status: 'TIMEOUT',
          time: this.timeout,
          ip
        };
      }
      return {
        status: 'MALFORMED',
        time: 0,
        ip
      };
    }
  }

  private parsePingTime(output: string, isWindows: boolean): number | null {
    if (isWindows) {
      // Try multiple patterns for Windows ping output
      // Format: "time=123ms" or "time<123ms" or "time<=123ms"
      let match = output.match(/time[<=](\d+)ms/i);
      if (match) {
        return parseFloat(match[1]) / 1000.0;
      }
      // Alternative: "Reply from ... time=123ms"
      match = output.match(/time=(\d+)ms/i);
      if (match) {
        return parseFloat(match[1]) / 1000.0;
      }
    } else {
      // Linux ping output: "time=12.345 ms" or "time=12.345ms"
      // Try the standard format first
      let match = output.match(/time=([\d.]+)\s*ms/i);
      if (match) {
        return parseFloat(match[1]) / 1000.0;
      }
      // Try without space: "time=12.345ms"
      match = output.match(/time=([\d.]+)ms/i);
      if (match) {
        return parseFloat(match[1]) / 1000.0;
      }
      // Try alternative format from some ping implementations
      match = output.match(/(\d+\.?\d*)\s*ms/i);
      if (match) {
        return parseFloat(match[1]) / 1000.0;
      }
    }
    return null;
  }

  // Remove pingFromNodes - not needed for agent
}

File: services/ping_service.py
"""
Ping Service - Perform ping checks from multiple nodes
"""
import asyncio
import subprocess
import platform
from typing import List, Dict, Optional
from models.schemas import PingResult, NodeInfo
from utils.helpers import is_valid_ip, parse_host_port


class PingService:
    """Service for performing ping checks"""
    
    def __init__(self):
        self.ping_count = 4  # Number of pings to perform
        self.timeout = 3  # Timeout in seconds
    
    async def ping_host(self, host: str, node: NodeInfo) -> List[List]:
        """Ping host from a specific node"""
        hostname, port = parse_host_port(host)
        
        # Resolve hostname to IP if needed
        ip = await self._resolve_hostname(hostname)
        if not ip:
            return [[None]]  # Unable to resolve
        
        # Perform ping
        results = []
        for _ in range(self.ping_count):
            result = await self._ping_ip(ip)
            results.append(result)
        
        return [results]
    
    async def _resolve_hostname(self, hostname: str) -> Optional[str]:
        """Resolve hostname to IP address"""
        if is_valid_ip(hostname):
            return hostname
        
        try:
            import socket
            loop = asyncio.get_event_loop()
            ip = await loop.run_in_executor(None, socket.gethostbyname, hostname)
            return ip
        except Exception:
            return None
    
    async def _ping_ip(self, ip: str) -> List:
        """Ping a single IP address"""
        try:
            # Determine ping command based on OS
            system = platform.system().lower()
            if system == "windows":
                cmd = ["ping", "-n", "1", "-w", str(self.timeout * 1000), ip]
            else:
                cmd = ["ping", "-c", "1", "-W", str(self.timeout), ip]
            
            # Run ping command
            process = await asyncio.create_subprocess_exec(
                *cmd,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE
            )
            
            stdout, stderr = await asyncio.wait_for(
                process.communicate(),
                timeout=self.timeout + 1
            )
            
            if process.returncode == 0:
                # Parse response time from output
                output = stdout.decode()
                time = self._parse_ping_time(output, system)
                if time is not None:
                    return ["OK", time, ip]
                else:
                    return ["OK", 0.0, ip]
            else:
                return ["TIMEOUT", self.timeout]
        
        except asyncio.TimeoutError:
            return ["TIMEOUT", self.timeout]
        except Exception as e:
            return ["MALFORMED", 0.0]
    
    def _parse_ping_time(self, output: str, system: str) -> Optional[float]:
        """Parse ping time from output"""
        import re
        
        if system == "windows":
            # Windows format: time=123ms or time<1ms
            match = re.search(r'time[<=](\d+)ms', output, re.IGNORECASE)
            if match:
                return float(match.group(1)) / 1000.0
        else:
            # Linux/Mac format: time=123.456 ms
            match = re.search(r'time=([\d.]+)\s*ms', output, re.IGNORECASE)
            if match:
                return float(match.group(1)) / 1000.0
        
        return None
    
    async def ping_from_nodes(self, host: str, nodes: Dict[str, NodeInfo]) -> Dict[str, List[List]]:
        """Ping host from multiple nodes"""
        tasks = {
            node_id: self.ping_host(host, node)
            for node_id, node in nodes.items()
        }
        
        results = await asyncio.gather(*tasks.values(), return_exceptions=True)
        
        return {
            f"{node_id}.node.check-host.net": result
            if not isinstance(result, Exception) else [[None]]
            for node_id, result in zip(tasks.keys(), results)
        }

